apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PgNumeric.cs

```diff
@@ -0,0 +1,487 @@
+        private static readonly BigInteger s_maxValueWithoutScale = BigInteger.Pow(10, MaxPrecision) - 1;
+
+        private static readonly BigInteger s_maxValueWithScale = ((BigInteger.Pow(10, MaxDigitsBeforeDecimalPoint) - 1) / (BigInteger.Pow(10, MaxScale)));
+
+        private static readonly BigInteger s_minValueWithScale = -s_maxValueWithScale;
+
+        private static readonly BigInteger s_minValueWithoutScale = -s_maxValueWithoutScale;
+
+        private static readonly Dictionary<int, BigInteger> s_powersOf10 = new Dictionary<int, BigInteger>();
+
+        // TODO: Don't require a 0 before the decimal point.
+        // TODO: Replace with manual validation if we find this is a performance bottleneck (as it was with BigQuery field name validations see #4976).
+        private static readonly Regex s_validation = new Regex(@"^-?[0-9]+\.?[0-9]*$", RegexOptions.Compiled);
+
+        // Note: the following properties must be declared *after* s_maxValue and s_minValue. Initialization order matters.
+
+        /// <summary>
+        /// Zero represented as a <see cref="PgNumeric"/>. This is the default value for the type.
+        /// </summary>
+        public static PgNumeric Zero { get; } = default;
+
+        /// <summary>
+        /// NaN (Not a number) represented as a <see cref="PgNumeric"/>. 
+        /// </summary>
+        public static PgNumeric NaN { get; } = Parse(NotANumber);
+
+        /// <summary>
+        /// The maximum valid value for <see cref="PgNumeric"/>, equal to BigInteger.Pow(10, MaxPrecision) - 1
+        /// </summary>
+        public static PgNumeric MaxValue { get; } = new PgNumeric(s_maxValueWithoutScale, 0, false);
+
+        /// <summary>
+        /// The minimum valid value for <see cref="PgNumeric"/>, equal to -(BigInteger.Pow(10, MaxPrecision) - 1)
+        /// </summary>
+        public static PgNumeric MinValue { get; } = new PgNumeric(s_minValueWithoutScale, 0, false);
+
+        // The scaled value of PgNumeric would be stored in this variable.
+        private readonly BigInteger _scaledValue;
+
+        // This flag would be true if PgNumeric is not a number.
+        private readonly bool _notANumber;
+
+        // The scale of numeric value stored.
+        private readonly int _scale;
+
+        private PgNumeric(BigInteger integer, int scale, bool isNotANumber)
+        {
+            if (!IsRawValueInRange(integer, scale))
+            {
+                throw new OverflowException("PgNumeric value out of range");
+            }
+
+            if (isNotANumber && integer != default && scale != default)
+            {
+                throw new ArgumentException("Invalid value and scale provided to PgNumeric for NaN");
+            }
+
+            _scale = scale;
+            _notANumber = isNotANumber;
+            _scaledValue = integer;
+        }
+
+        private static bool IsRawValueInRange(BigInteger integer, int scale)
```

Comment:
It's not clear to me that this is correct. I think we need to talk about it more.

Reply:
Even now I've seen the chat comment about a maximum number of decimal digits, I'm not sure this is actually correct. It would be quite odd for the maximum values to be a step function like this rather than something that changes by a factor of 10 per change of 1 in scale.